solo-io/gloo
# Refactor Kubernetes E2E test definition for ease of use

## Description

Introduces abstractions over the test cases for Ordered and Unordered sets of tests within a suite. This allows us to define the run behavior of the tests separate from the suite that runs them.

Also refactors the tests out into an importable `tests` package (`*_test` packages cannot be imported). This will allow us to run the tests in the same way in EE, and also make sure the test suites in OSS and EE stay in lockstep.

Any tests which did not have a comment about needing to be ordered were assumed to be unordered. There is a non-zero risk that this introduces some flakes if the test order was in fact important.

## Context

Tests added to these suites in OSS had to be added in the same manner in EE. By exporting the entire pack of tests to run, the EE suites can import them to run on their bespoke `e2e.TestInstallation`.

## Code changes
- New types and interface to abstract the concept of Ordered and Unordered groups of tests
- Tests actually run in the suite were separated from the suite. This adds a little unnecessary code distance, but in order to import them they could not be in a `*_test` package and it seemed cleaner to me (and possibly necessary for testify but not sure) to leave the suites in `*_test` package.

## Testing steps

I manually verified behavior by opening https://github.com/solo-io/solo-projects/pull/6314 and running nightlies off of that branch to see that the expected suite ran (at that time just the k8s_gw_tests had been migrated).

## Interesting decisions

In order to reduce verbosity of defining a new test, we strip back some of the injection to the actual `Run` function. This means that we cannot need to reference the three things we need injected (`ctx, t, testInstallation`) when we are registering tests. To reconcile with this, what is registered is actually the feature suite construction function. We have enforced this by defining a new typedef for that signature and making a compiler assertion at the top of each feature suite for its construction function.

A side-effect of this is that nested test structures such as used for Glooctl in `k8s_gw_test.go` no longer work since we don't have access to the injected values at registration time. To resolve this, nested test structures must be defined as suites. An example of this can be found in the same Glooctl test, now located in `k8s_gw_tests.go`.

# Checklist:
